chanfitness/chaneventstore.go

```diff
@@ -0,0 +1,381 @@
+	// events.
+	peers map[[33]byte]bool
+
+	// lifespanRequests serves requests for the lifespan of channels.
+	lifespanRequests chan lifespanRequest
+
+	// uptimeRequests serves requests for the uptime of channels.
+	uptimeRequests chan uptimeRequest
+
+	quit chan struct{}
+
+	wg sync.WaitGroup
+}
+
+// Config provides the event store with functions required to monitor channel
+// activity. All elements of the config must be non-nil for the event store to
+// operate.
+type Config struct {
+	// Channel events provides a subscription client which provides a
+	// stream of channel events.
+	ChannelEvents func() (*subscribe.Client, error)
+
+	// PeerEvents provides a subscription client which provides a stream of
+	// peer online/offline events.
+	PeerEvents func() (*subscribe.Client, error)
+
+	// Get open channels provides a list of existing open channels
+	// which is used to populate the ChannelEventStore with a set of channels
+	// on startup.
+	GetOpenChannels func() ([]*channeldb.OpenChannel, error)
+}
+
+// lifespanRequest contains the channel ID required to query the store for a
+// chanel's lifespan and a blocking response channel on which the result is sent.
+type lifespanRequest struct {
+	channelID    uint64
+	responseChan chan lifespanResponse
+}
+
+// lifespanResponse contains the response to a lifespanRequest and an error if
+// one occurred.
+type lifespanResponse struct {
+	start time.Time
+	end   time.Time
+	err   error
+}
+
+// uptimeRequest contains the parameters required to query the store for a
+// channel's uptime and a blocking response channel on which the result is sent.
+type uptimeRequest struct {
+	channelID    uint64
+	startTime    time.Time
+	endTime      time.Time
+	responseChan chan uptimeResponse
+}
+
+// uptimeResponse contains the response to an uptimeRequest and an error if one
+// occurred.
+type uptimeResponse struct {
+	uptime time.Duration
+	err    error
+}
+
+// NewChannelEventStore initializes an event store with the config provided.
+// Note that this function does not start the main event loop, Start() must be
+// called.
+func NewChannelEventStore(config *Config) *ChannelEventStore {
+	store := &ChannelEventStore{
+		cfg:              config,
+		channels:         make(map[uint64]*chanEventLog),
+		peers:            make(map[[33]byte]bool),
+		lifespanRequests: make(chan lifespanRequest),
+		uptimeRequests:   make(chan uptimeRequest),
+		quit:             make(chan struct{}),
+	}
+
+	return store
+}
+
+// Start adds all existing open channels to the event store and starts the main
+// loop which records channel and peer events, and serves requests for
+// information from the store.
+func (c *ChannelEventStore) Start() error {
+	// Create a subscription to channel events.
+	channelClient, err := c.cfg.ChannelEvents()
+	if err != nil {
+		return err
+	}
+
+	// Create a subscription to peer events.
+	peerClient, err := c.cfg.PeerEvents()
+	if err != nil {
+		return err
```

Comment:
should cancel `channelClient` before returning here. Same with peerClient in error case below.